golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,270 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+	vstoragetypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	snapshots "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+// SnapshotFormat 1 is a proto message containing an artifact name, and the binary artifact data
+const SnapshotFormat = 1
+
+type activeExtensionSnapshot struct {
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// The function to retrieve the exported artifacts
+	retrieveSwingStoreExport func() error
+	// The writer for extensions payloads
+	payloadWriter snapshots.ExtensionPayloadWriter
+}
+
+// SwingsetSnapshotter manages Swingset snapshots, ensuring insensitivity to
+// sub-block timing of communication with the JS side, and enforcing
+// concurrency requirements.
+// The caller of this submodule must arrange block level commit synchronization,
+// to ensure the results are deterministic.
+type SwingsetSnapshotter struct {
+	isConfigured            func() bool
+	takeAppSnapshot         func(height int64)
+	newRestoreContext       func(height int64) sdk.Context
+	swingStoreSnapshotter   *SwingStoreSnapshotter
+	getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry
+	logger                  log.Logger
+	activeExtensionSnapshot *activeExtensionSnapshot
+}
+
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
+	return SwingsetSnapshotter{
+		isConfigured:    func() bool { return app.SnapshotManager() != nil },
+		takeAppSnapshot: app.Snapshot,
+		newRestoreContext: func(height int64) sdk.Context {
+			return app.NewUncachedContext(false, tmproto.Header{Height: height})
+		},
+		logger:                  app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "extension snapshotter"),
+		swingStoreSnapshotter:   swingStoreSnapshotter,
+		getSwingStoreExportData: getSwingStoreExportData,
+		activeExtensionSnapshot: nil,
+	}
+}
+
+// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+// Implements ExtensionSnapshotter
+func (snapshotter *SwingsetSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+// SnapshotFormat returns the default format the extension snapshotter uses to encode the
+// payloads when taking a snapshot.
+// It's defined within the extension, different from the global format for the whole state-sync snapshot.
```

Comment:
I don't think I'm clear on the various levels of versioning, or e.g. if the value returned from SnapshotFormat corresponds with a single artifact or with the collection of all artifacts. Is it possible to improve this a bit?

Reply:
This is inherited by the cosmos-sdk. Versioning applies to all "extensions payloads of this module" I believe.